linked_list/linked_list.py

```diff
@@ -13,63 +13,144 @@ def __init__(self):
+            self.head= new_node
+        else:
+            new_node = Node(value=value, next_node=self.head)
+            self.head = new_node
+
 
     # method to find if the linked list contains a node with specified value
     # returns true if found, false otherwise
     # Time Complexity: ?
     # Space Complexity: ?
     def search(self, value):
-        pass
+        if self.head == None:
+            return False
+        else:
+            current = self.head
+            while current != None:
+                if current.value == value:
+                    return True
+                else:
+                    current = current.next
+            return False
+        
 
     # method that returns the length of the singly linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def length(self):
-        pass
+        if self.head == None:
+            return 0
+        else:
+            counter = 0
+            current = self.head
+            while current != None:
+                counter += 1
+                current = current.next
+            return counter 
+
 
     # method that returns the value at a given index in the linked list
     # index count starts at 0
     # returns None if there are fewer nodes in the linked list than the index value
     # Time Complexity: ?
     # Space Complexity: ?
     def get_at_index(self, index):
-        pass
+        if self.head == None:
+            return None
+        else:
+            counter = 0 
+            current = self.head
+            while current != None:
+                if counter == index:
+                    return current.value
+                else:
+                    counter += 1
+                    current = current.next
+            return None
 
     # method that returns the value of the last node in the linked list
     # returns None if the linked list is empty
     # Time Complexity: ?
     # Space Complexity: ?
     def get_last(self):
-        pass
+        if self.head == None:
+            return None
+        else:
+            current = self.head
+            while current != None:
+                if current.next == None:
+                    return current.value
+                else:
+                    current = current.next
 
     # method that inserts a given value as a new last node in the linked list
     # Time Complexity: ?
     # Space Complexity: ?
     def add_last(self, value):
-        pass
+        new_node = Node(value, next_node=None)
+        if self.head == None:
+            self.head = new_node
+        else:
+            current = self.head
+            while current != None:
+                if current.next == None:
+                    new_node = Node(value, next_node=None)
+                    current.next = new_node
+                    current = current.next
+                    return 
+                else:
+                    current = current.next

```

Comment:
like above, we could change up the while statement in order to use less lines of code:

```suggestion
            while current.next:
                current = current.next
            current.next = Node(value, next_node=None)
```